lib/cache/plugins.go

```diff
@@ -0,0 +1,190 @@
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return item, nil
+	}
+
+	// Cache is enabled; load plugins from cache store.
+	// Depending on the watcher configuration, the cached plugin may or may not contain secrets.
+	// The returned plugin will be stripped of secrets if requested.
+	plugin, err := rg.store.get(pluginNameIndex, name)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return stripAndClonePluginSecrets(plugin, withSecrets), nil
+}
+
+// GetPlugins returns a full list of plugins from the cache.
+// Falls back to backend API if caching is disabled.
+// The `withSecrets` flag controls whether secrets are included in the response.
+func (c *Cache) GetPlugins(ctx context.Context, withSecrets bool) ([]types.Plugin, error) {
+	_, span := c.Tracer.Start(ctx, "cache/GetPlugins")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.plugins)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		// Cache is disabled; fetch all plugins from backend.
+		// The backend will filter secrets based on withSecrets.
+		plugins, err := c.Config.Plugin.GetPlugins(ctx, withSecrets)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return plugins, nil
+	}
+
+	// Cache is enabled; load plugins from cache store.
+	// Strip secrets if requested.
+	plugins := make([]types.Plugin, 0, rg.store.len())
+	for item := range rg.store.resources(pluginNameIndex, "", "") {
+		plugins = append(plugins, stripAndClonePluginSecrets(item, withSecrets))
+	}
+	return plugins, nil
+}
+
+// ListPlugins returns a paginated list of plugins from the cache or backend.
+// If caching is disabled, it uses the backend directly. Otherwise, it slices from in-memory cache.
+// The `withSecrets` flag controls inclusion of secrets in the result.
+func (c *Cache) ListPlugins(ctx context.Context, limit int, startKey string, withSecrets bool) ([]types.Plugin, string, error) {
+	_, span := c.Tracer.Start(ctx, "cache/ListPlugins")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.plugins)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		// Cache is disabled; paginate from backend directly.
+		// The backend applies secret filtering based on withSecrets.
+		items, nextKey, err := c.Config.Plugin.ListPlugins(ctx, limit, startKey, withSecrets)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		return items, nextKey, nil
+	}
+
+	// Cache is enabled; load plugins from cache store.
+	// Limit the number of items returned to the page size.
+	if limit <= 0 {
+		limit = apidefaults.DefaultChunkSize
+	}
+
+	var plugins []types.Plugin
+	var nextKey string
+	for item := range rg.store.resources(pluginNameIndex, startKey, "") {
+		plugins = append(plugins, stripAndClonePluginSecrets(item, withSecrets))
+		if len(plugins) == limit {
+			nextKey = backend.GetPaginationKey(item)
+			break
+		}
```

Comment:
This should be the other way around, right? Do we not have a test that catches this? The start key is included, so `nextKey` should be calculated with the item after the last one.

```suggestion
		if len(plugins) >= limit {
			nextKey = backend.GetPaginationKey(item)
			break
		}
		plugins = append(plugins, stripAndClonePluginSecrets(item, withSecrets))
```

Reply:
Fixed. 

Added pagination test is a seperate PR:  https://github.com/gravitational/teleport/pull/56547